src/iceberg/util/projection_util_internal.h

```diff
@@ -0,0 +1,499 @@
+#include "iceberg/transform_function.h"
+#include "iceberg/util/checked_cast.h"
+#include "iceberg/util/decimal.h"
+#include "iceberg/util/macros.h"
+
+namespace iceberg {
+
+class ProjectionUtil {
+ private:
+  static Result<std::shared_ptr<UnboundPredicate>> TransformSet(
+      std::string_view name, const std::shared_ptr<BoundSetPredicate>& predicate,
+      const std::shared_ptr<TransformFunction>& func) {
+    std::vector<Literal> transformed;
+    transformed.reserve(predicate->literal_set().size());
+    for (const auto& lit : predicate->literal_set()) {
+      ICEBERG_ASSIGN_OR_RAISE(auto transformed_lit, func->Transform(lit));
+      transformed.push_back(std::move(transformed_lit));
+    }
+    return Expressions::Predicate(predicate->op(), std::string(name),
+                                  std::move(transformed));
+  }
+
+  static Result<std::shared_ptr<UnboundPredicate>> TruncateArray(
+      std::string_view name, const std::shared_ptr<BoundLiteralPredicate>& predicate,
+      const std::shared_ptr<TransformFunction>& func) {
+    const Literal& boundary = predicate->literal();
+
+    ICEBERG_ASSIGN_OR_RAISE(auto transformed, func->Transform(boundary));
+
+    switch (predicate->op()) {
+      case Expression::Operation::kLt:
+      case Expression::Operation::kLtEq:
+        return Expressions::Predicate(Expression::Operation::kLtEq, std::string(name),
+                                      transformed);
+      case Expression::Operation::kGt:
+      case Expression::Operation::kGtEq:
+        return Expressions::Predicate(Expression::Operation::kGtEq, std::string(name),
+                                      transformed);
+      case Expression::Operation::kEq:
+        return Expressions::Predicate(Expression::Operation::kEq, std::string(name),
+                                      transformed);
+      case Expression::Operation::kStartsWith:
+        return Expressions::Predicate(Expression::Operation::kStartsWith,
+                                      std::string(name), transformed);
+      default:
+        return nullptr;
+    }
+    std::unreachable();
+  }
+
+  template <typename T>
+    requires std::is_same_v<T, int32_t> || std::is_same_v<T, int64_t>
+  static Result<std::shared_ptr<UnboundPredicate>> TruncateInteger(
+      std::string_view name, const std::shared_ptr<BoundLiteralPredicate>& predicate,
+      const std::shared_ptr<TransformFunction>& func) {
+    const Literal& literal = predicate->literal();
+
+    switch (predicate->op()) {
+      case Expression::Operation::kLt: {
+        // adjust closed and then transform ltEq
+        if constexpr (std::is_same_v<T, int32_t>) {
+          ICEBERG_ASSIGN_OR_RAISE(
+              auto transformed,
+              func->Transform(literal.type()->type_id() == TypeId::kDate
+                                  ? Literal::Date(std::get<T>(literal.value()) - 1)
+                                  : Literal::Int(std::get<T>(literal.value()) - 1)));
```

Comment:
This does not look safe since `predicate` may be of any type but `TruncateInteger` indicates that only integer variants are accepted.